### PR TITLE
Fixes #24532 - Allow changing display type and keyboard in oVirt

### PR DESCRIPTION
**Pull request by UXabre**
Depends on https://github.com/fog/fog-ovirt/pull/47

### CODE REVIEW

File: app/controllers/api/v2/compute_resources_controller.rb [hunk 1]
```diff
@@ -44,7 +44,8 @@ def show
           param :domain, String, :desc => N_("for OpenStack only")
           param :server, String, :desc => N_("for VMware")
           param :set_console_password, :bool, :desc => N_("for Libvirt and VMware only")
-          param :display_type, %w(VNC SPICE), :desc => N_('for Libvirt only')
+          param :display_type, %w(VNC SPICE), :desc => N_('for Libvirt and oVirt only')
+          param :keyboard, %w(ar de-ch es fo fr-ca hu ja mk no pt-br sv da en-gb et fr fr-ch is lt nl pl ru th de en-us fi fr-be hr it lv nl-be pt sl tr), :desc => N_('for oVirt only')
```
Comment (timogoebel) [inline app/controllers/api/v2/compute_resources_controller.rb]: Can you use `ALLOWED_KEYBOARD_LAYOUTS` here?

**Comment by UXabre** [comment]
- I've added a few tests for the keyboard layout as well
- Fixed the hound issues
- Greatly decreased the needed changes on fog-ovirt side

File: app/controllers/api/v2/compute_resources_controller.rb [hunk 2]
```diff
@@ -44,7 +44,8 @@ def show
           param :domain, String, :desc => N_("for OpenStack only")
           param :server, String, :desc => N_("for VMware")
           param :set_console_password, :bool, :desc => N_("for Libvirt and VMware only")
-          param :display_type, %w(VNC SPICE), :desc => N_('for Libvirt only')
+          param :display_type, %w(VNC SPICE), :desc => N_('for Libvirt and oVirt only')
+          param :keyboard, Foreman::Model::Ovirt::ALLOWED_KEYBOARD_LAYOUTS, :desc => N_('for oVirt only')
```
Comment (timogoebel) [inline app/controllers/api/v2/compute_resources_controller.rb]: `Foreman::Model::Ovirt` might actually not be available in a prod environment (it's an optional bundler group). Maybe we can move this to compute resource model.

Comment (UXabre) [reply app/controllers/api/v2/compute_resources_controller.rb]: Oh, yeah, didn't think about that one! Will see to move this

Comment (UXabre) [reply app/controllers/api/v2/compute_resources_controller.rb]: On the other hand, would it not be presumptuous to say that all keyboard layouts are always the same for all compute resources? And thus, would it not be easier to just set the type to String instead (and let each compute resource for itself decide if the layout is valid?)

Comment (timogoebel) [reply app/controllers/api/v2/compute_resources_controller.rb]: Without wanting to be overly strict, I think we should make sure the possible values can be found in API docs. But I agree we should not assume the values are the same for all compute resources.

Comment (UXabre) [reply app/controllers/api/v2/compute_resources_controller.rb]: Also agreed that the documentation should reflect the possible values. Perhaps it's best to, for now, make the assumption on equality and only _cross the bridge_ when we get to a case in which keyboard layouts are different?

If OK for you, I'll change it to your first proposal :-)

Comment (timogoebel) [reply app/controllers/api/v2/compute_resources_controller.rb]: Sure, go ahead.

**Comment by UXabre** [comment]
I've noticed al lot of tests have failed but I don't think this is because of my change?
Was this a fluke? (earlier yesterday it seemed to build just fine though)

**Comment by shiramax** [comment]
[test foreman]

**Comment by shiramax** [comment]
[test katello]

**Comment by UXabre** [comment]
Hmm, I noticed the tests start failing again; although, at first, it doesn't seem to be related to my change, it does seem to make the ComputeResource tests unstable?

**Comment by UXabre** [comment]
Good news! I AM to blame! Will try and figure out what is going on...

**Comment by UXabre** [comment]
Yes, appears that replacing the keyboards in "compute_resources_controller.rb" with a variable, isn't possible. I tried a few things at least but seems literally copying the literal is the only way to make it work? Anyone experience with apipie for this?

**Comment by UXabre** [comment]
Yeah, brainfart solved! It must've been a hard day, as I tried to reference the constant in a completely wrong scope... fixed now! Awaiting the build-results...

**Comment by timogoebel** [comment]
Code looks good to me, but I'll pass this along to somebody who has oVirt knowledge.

**Review by shiramax — CHANGES_REQUESTED** [review]
@UXabre 
I've tested both patches:

-  I tried to create and host with vnc and and keyborad layout "lt" and I got the following error:
```
Failed to create a compute ovirt (oVirt) instance kelly-aquil.bla: Ovirt client returned an error: The string '' isn't a valid value for the 'DisplayType' enumerated type. Valid values are: 'vnc', 'spice'.
```

- As we agreed, the console shouldn't be selected in the compute resource, the reason for that is that if a user would like to use both options, he will need to create two compute resource: one for spice and one for vnc.
please move the option of the console to the creation of the host.

- please add the option "spice+vnc" for the console

**Comment by UXabre** [comment]
Hey @shiramax!

Thanks for reviewing!

1. Hmm, I'll have to check why it failed and I'll add a test to prevent this from happening in the future
2. My vote, consistency wise, would be to keep it at both places; but just like with the keyboard, let the compute-resource variant act as a default and allow host-specific override to something else. This to be fairly in line with libvirt CR.
3. Can do!

So basically; I'll fix all those points, but will probably be pushed to somewhere later this week

Kind regards!

**Comment by UXabre** [comment]
So, finally managed to make the changes. As said, I currently have a global default option in the compute resource but each VM can individually be configured to override this setting. It can even be changed if the VM is down.

As for the "VNC+Spice" option, I didn't find anything in the API that could help me configure such a thing, nor does setting this in the oVirt UI. All I could see was that Spice is set for such VMs. Perhaps this will need more work on the fog ovirt plugin to support that case?

Finally, tried to reproduce the bug you encountered but did not manage to do so; could you please give me the detailed steps of what you did please? :-)

**Review by shiramax — CHANGES_REQUESTED** [review]
@UXabre thanks!
- I wasn't able to reproduce the bug I saw in the first review too, so you can ignore that.
- In the compute resource,  If I choose a SPICE, the Keyboard layout is still presented, and I think it's a bit misleading, It should act the same as it acts in the VM form.
-about the spice+vnc, yes it would probably will need more support in fog-ovirt. From what I know, we should add both devices to the vm:
https://github.com/oVirt/ovirt-engine-sdk-ruby/blob/master/sdk/examples/add_vnc_console.rb

**Comment by UXabre** [comment]
Hi @shiramax ,

Thank you so much for further reviewing this PR.

 - Glad to hear the bug isn't encountered anymore :-)
 - I understand the confusion, at first I did it like you commented but I then realized these are actually global defaults; a global console protocol but also a global keyboard setting in case one **would** select VNC further down the line. Perhaps I have to annotate that this setting only applies to VNC display type or should I simply call it "Default VNC Keyboard Layout"? I hope I was able to communicate the reason for keeping it there fixed? But again, it's just a proposal from my end; need not be exactly that :-)
 - Would love to see this addition! But would rather love it as a separate issue as it's a bit out of my comfortzone and I fear that this PR would loose some velocity by increasing the scope. Also, as my current tests indiciate, oVirt always returns SPICE as being the display type for both VNC+SPICE as SPICE console VMs; this means that oVirt will always load the SPICE web-client for those VMs (which doesn't seem to work all too well in the version shipped with foreman, the newer version does a much better job but isn't 100% there yet). IMO this would be the work break-down:
    1. Update fog ovirt to include support for VNC+SPICE (via link you added in your comment it seems doable)
     2. Reflect this in the Foreman UI (meaning that console interpretation would probably have to change) by both showing that possible consoles but perhaps the console button should also change in a drop-down so a user could select which type to use as current console? Or we could keep the UI from this PR to select which type of console to open when clicking the console-button
      3. Lots of testing to validate these changes

Again, for me that kind of effort looks like it would be more efficiëntly split up? 

Hope to hear from you soon! :-)

**Comment by shiramax** [comment]
@UXabre 
- changing the name to 'Default VNC Keyboard Layout' is acceptable 
- NP, that could be in a separate issue

**Review by shiramax — CHANGES_REQUESTED** [review]
@UXabre

changing the name to 'Default VNC Keyboard Layout' is acceptable
NP, that could be in a separate issue

**Comment by UXabre** [comment]
Hey @shiramax , I've made the adjustement and thanks! Do you want to create the follow up feature-issue or should I create it? :-)

Happy reviewing!

Kr,
Arend

**Comment by shiramax** [comment]
[test katello]

**Comment by UXabre** [comment]
Seems the katello built is stuck :-(

**Comment by UXabre** [comment]
Hey, What should happen next? Should we try to build Katello again? :-)

**Comment by mmoll** [comment]
@UXabre please rebase

**Comment by UXabre** [comment]
Hi @mmoll , successfully rebased and pushed this PR :-) (awaiting all check results)

**Comment by mmoll** [comment]
all :green_heart:

**Review by shiramax — DISMISSED** [review]

**Review by mmoll — CHANGES_REQUESTED** [review]
We need to wait for a fog-ovirt-release with the needeed code included, thus blocking for now.

**Comment by UXabre** [comment]
Hey @mmoll I see a new release of fog-ovirt made the cut! Probably would need a PR somewhere first to include the changed package in Foreman? Or does it always take the latest greatest?

**Comment by mmoll** [comment]
@UXabre as 2de40ae went in, 1.2.0 is used now.

**Review by mmoll — APPROVED** [review]

**Comment by mmoll** [comment]
merged, thanks @UXabre!
